hypothesis-python/tests/numpy/test_gen_data.py

```diff
@@ -316,14 +316,24 @@ def test_overflowing_integers_are_deprecated(fill, data):
 
 
 @pytest.mark.parametrize('fill', [False, True])
+@pytest.mark.parametrize('dtype,strat', [
+    ('float16', st.floats(min_value=65520, allow_infinity=False)),
+    ('complex64', st.complex_numbers(10**300, allow_infinity=False)),
+    ('U1', st.text(min_size=2, max_size=2)),
+    ('S1', st.binary(min_size=2, max_size=2)),
+])
 @checks_deprecated_behaviour
-@given(st.data())
-def test_overflowing_floats_are_deprecated(fill, data):
-    kw = dict(elements=st.floats(min_value=65520, allow_infinity=False))
+@given(data=st.data())
+def test_unrepresentable_elements_are_deprecated(fill, dtype, strat, data):
     if fill:
-        kw = dict(elements=st.nothing(), fill=kw['elements'])
-    arr = data.draw(nps.arrays(dtype='float16', shape=(1,), **kw))
-    assert np.isinf(arr[0])
+        kw = dict(elements=st.nothing(), fill=strat)
+    else:
+        kw = dict(elements=strat)
+    arr = data.draw(nps.arrays(dtype=dtype, shape=(1,), **kw))
+    try:
+        assert np.isinf(arr[0])
+    except TypeError:
+        assert len(arr[0]) <= 1
```

Comment:
I'm very confused about what's going on here. I don't see how this assertion could ever pass - either the value is not something with a len, or we generated it with `min_size=2`.

Reply:
I will add a comment - `np.isinf` raises a `TypeError` when passed a (unicode or byte) string.  In turn, the string was generated by Hypothesis with length two, *but* because the dtype of the array is for length-one strings it must have one character (or zero, if it was of the form `"\0a"`).  This confusing behaviour is exactly what our new code aims to detect and warn about.